File: launcher/src/main/java/com/droibit/android/customtabs/launcher/CustomTabsLauncher.java
package com.droibit.android.customtabs.launcher;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.RestrictTo;
import android.support.annotation.Size;
import android.support.customtabs.CustomTabsIntent;
import java.util.List;

import static android.support.annotation.RestrictTo.Scope.LIBRARY;
import static com.droibit.android.customtabs.launcher.ChromePackage.CHROME_PACKAGES;

/**
 * A utility class that directly launches a browser that supports Custom Tabs.<br/>
 * <p>
 * This library try to launch Chrome(s) as the default behavior,
 * but if they are not installed you can launch a browser that supports other Custom Tabs using {@link LaunchNonChromeCustomTabsFallback}.
 * </p>
 * <h3>Basic usage:</h3>
 * <pre>{@code
 *  final CustomTabsIntent customTabsIntent = ...;
 *  CustomTabsLauncher.launch(ctx, customTabsIntent, URI);
 * }</pre>
 * <h3>Launch non-Chrome browser as Fallback</h3>
 * <pre>{@code
 *  final CustomTabsIntent customTabsIntent = ...;
 *  final exampleNonChromePackages = Arrays.asList(
 *    "org.mozilla.firefox",
 *    "com.microsoft.emmx",
 *  );
 *  CustomTabsLauncher.launch(ctx, customTabsIntent, URI, exampleNonChromePackages);
 * }</pre>
 */
public final class CustomTabsLauncher {

  /**
   * Fallback for launch a non-Chrome browser that supports Custom Tabs.
   */
  public static class LaunchNonChromeCustomTabsFallback implements CustomTabsFallback {

    private final List<String> customTabsPackages;

    /**
     * @param customTabsPackages Package list of non-Chrome browsers supporting Custom Tabs. The top of the list is used with the highest priority.
     */
    public LaunchNonChromeCustomTabsFallback(
        @NonNull @Size(min = 1) List<String> customTabsPackages) {
      this.customTabsPackages = customTabsPackages;
    }

    @Override public void openUrl(
        @NonNull Context context,
        @NonNull Uri uri,
        @NonNull CustomTabsIntent customTabsIntent) {
      IMPL.launch(context, customTabsIntent, uri, customTabsPackages, (c, u, i) -> {
        final Intent intent = new Intent(Intent.ACTION_VIEW, uri)
            .setFlags(customTabsIntent.intent.getFlags());
        context.startActivity(intent);
      });
    }
  }

  @RestrictTo(LIBRARY)
  private static final CustomTabsLauncherImpl IMPL = new CustomTabsLauncherImpl();

  /**
   * Opens the URL on a Custom Tabs if possible.
   *
   * @param context The source Context
   * @param customTabsIntent a CustomTabsIntent to be used if Custom Tabs is available
   * @param uri the Uri to be opened.
   */
  public static void launch(
      @NonNull Context context,
      @NonNull CustomTabsIntent customTabsIntent,
      @NonNull Uri uri) {
    launch(context, customTabsIntent, uri, null);
  }

  /**
   * Opens the URL on a Custom Tabs if possible. Otherwise fallsback to opening it on a WebView.
   *
   * @param context The source Context
   * @param customTabsIntent a CustomTabsIntent to be used if Custom Tabs is available
   * @param uri the Uri to be opened
   * @param fallback a {@link CustomTabsFallback} to be used if Custom Tabs is not available.
   */
  public static void launch(
      @NonNull Context context,
      @NonNull CustomTabsIntent customTabsIntent,
      @NonNull Uri uri,
      @Nullable CustomTabsFallback fallback) {
    IMPL.launch(context, customTabsIntent, uri, CHROME_PACKAGES, fallback);
  }
}

File: launcher/src/main/java/com/droibit/android/customtabs/launcher/CustomTabsLauncher.kt
package com.droibit.android.customtabs.launcher

import android.content.Context
import android.net.Uri
import androidx.browser.customtabs.CustomTabsIntent
import com.droibit.android.customtabs.launcher.CustomTabsLauncher.LaunchNonChromeCustomTabs
import com.droibit.android.customtabs.launcher.internal.CustomTabsLauncherImpl
import com.droibit.android.customtabs.launcher.CustomTabsPackage.CHROME_PACKAGES

/**
 * A utility class that directly launches a browser that supports Custom Tabs.
 *
 * This library try to launch Chrome(s) as the default behavior,
 * but if they are not installed you can launch a browser that supports other Custom Tabs using [LaunchNonChromeCustomTabs].
 *
 * * Basic usage:
 * ```
 * val customTabsIntent = ...;
 * CustomTabsLauncher.launch(ctx, customTabsIntent, URI);
 * ```
 *
 * * Launch non-Chrome browser as Fallback
 * ```
 * val customTabsIntent = ...;
 * CustomTabsLauncher.launch(ctx, customTabsIntent, URI, LaunchNonChromeCustomTabs(ctx));
 * ```
 */
object CustomTabsLauncher {
    private val IMPL = CustomTabsLauncherImpl()

    /**
     * Attempt to open the URL directly in a browser that supports Custom Tabs.
     *
     * @param context          The source Context
     * @param customTabsIntent a CustomTabsIntent to be used if Custom Tabs is available
     * @param uri              the Uri to be opened.
     * @param fallback         a [CustomTabsFallback] to be used if Custom Tabs is not available.
     */
    @Deprecated("Migrate to 'CustomTabsIntent.ensureChromeCustomTabsPackage' and 'CustomTabsIntent.launchUrl'.")
    @JvmStatic
    @JvmOverloads
    fun launch(
        context: Context,
        customTabsIntent: CustomTabsIntent,
        uri: Uri,
        fallback: CustomTabsFallback? = null
    ) {
        IMPL.launch(context, customTabsIntent, uri, CHROME_PACKAGES, fallback)
    }

    /**
     * Fallback for launch a non-Chrome browser that supports Custom Tabs.
     *
     * @param customTabsPackages Package list of non-Chrome browsers supporting Custom Tabs. The top of the list is used with the highest priority.
     */
    @Deprecated("Migrate to 'NonChromeCustomTabs'.")
    class LaunchNonChromeCustomTabs(
        private val customTabsPackages: List<String>
    ) : CustomTabsFallback {

        constructor(context: Context) : this(
            CustomTabsPackage.getNonChromeCustomTabsPackages(context)
        )

        override fun openUrl(
            context: Context,
            uri: Uri,
            customTabsIntent: CustomTabsIntent
        ) {
            if (customTabsPackages.isEmpty()) {
                customTabsIntent.launchUrl(context, uri)
            } else {
                IMPL.launch(context, customTabsIntent, uri, customTabsPackages, null)
            }
        }
    }
}

/**
 * To be used as a fallback to open the Uri when Custom Tabs is not available.
 */
@Deprecated("Migrate to 'CustomTabsPackageFallback'.")
fun interface CustomTabsFallback {
    /**
     * @param context          The source Context
     * @param uri              The Uri to load in the Custom Tab
     * @param customTabsIntent a source CustomTabsIntent
     */
    fun openUrl(context: Context, uri: Uri, customTabsIntent: CustomTabsIntent)
}